ctapipe/io/hdf5merger.py

```diff
@@ -0,0 +1,467 @@
+
+import tables
+from astropy.time import Time
+
+from ..core import Component, Provenance, traits
+from ..instrument.optics import FocalLengthKind
+from ..instrument.subarray import SubarrayDescription
+from ..utils.arrays import recarray_drop_columns
+from . import metadata
+from .hdf5tableio import DEFAULT_FILTERS
+
+
+class NodeType(enum.Enum):
+    # a single table
+    TABLE = enum.auto()
+    # a group comprising tel_XXX tables
+    TEL_GROUP = enum.auto()
+    # a group with children of the form /<property group>/<algorithm table>
+    ALGORITHM_GROUP = enum.auto()
+    # a group with children of the form /<property group>/<algorithm group>/<tel_XXX table>
+    ALGORITHM_TEL_GROUP = enum.auto()
+
+
+#: nodes to check for merge-ability
+_NODES_TO_CHECK = {
+    "/configuration/observation/scheduling_block": NodeType.TABLE,
+    "/configuration/observation/observation_block": NodeType.TABLE,
+    "/configuration/simulation/run": NodeType.TABLE,
+    "/simulation/service/shower_distribution": NodeType.TABLE,
+    "/simulation/event/subarray/shower": NodeType.TABLE,
+    "/simulation/event/telescope/impact": NodeType.TEL_GROUP,
+    "/simulation/event/telescope/images": NodeType.TEL_GROUP,
+    "/simulation/event/telescope/parameters": NodeType.TEL_GROUP,
+    "/dl1/event/subarray/trigger": NodeType.TABLE,
+    "/dl1/event/telescope/trigger": NodeType.TABLE,
+    "/dl1/event/telescope/images": NodeType.TEL_GROUP,
+    "/dl1/event/telescope/parameters": NodeType.TEL_GROUP,
+    "/dl2/event/telescope": NodeType.ALGORITHM_TEL_GROUP,
+    "/dl2/event/subarray": NodeType.ALGORITHM_GROUP,
+    "/dl1/monitoring/subarray/pointing": NodeType.TABLE,
+    "/dl1/monitoring/telescope/pointing": NodeType.TEL_GROUP,
+}
+
+
+def _get_required_nodes(h5file):
+    """Return nodes to be required in a new file for appending to ``h5file``"""
+    required_nodes = set()
+    for node, node_type in _NODES_TO_CHECK.items():
+        if node not in h5file.root:
+            continue
+
+        if node_type in (NodeType.TABLE, NodeType.TEL_GROUP):
+            required_nodes.add(node)
+
+        elif node_type is NodeType.ALGORITHM_GROUP:
+            for kind_group in h5file.root[node]._f_iter_nodes("Group"):
+                for table in kind_group._f_iter_nodes("Table"):
+                    required_nodes.add(table._v_pathname)
+
+        elif node_type is NodeType.ALGORITHM_TEL_GROUP:
+            for kind_group in h5file.root[node]._f_iter_nodes("Group"):
+                for algorithm_group in kind_group._f_iter_nodes("Group"):
+                    required_nodes.add(algorithm_group._v_pathname)
+        else:
+            raise ValueError(f"Unhandled node type: {node_type} of {node}")
+
+    return required_nodes
+
+
+class CannotMerge(IOError):
+    """Raised when trying to merge incompatible files"""
+
+
+def split_h5path(path):
+    """
+    Split a path inside an hdf5 file into parent / child
+    """
+    head, _, tail = path.rstrip("/").rpartition("/")
```

Comment:
Why would there be a trailing `/`?

Optionally: `head, tail = path.rsplit("/", 1)`, but his is fine as well.

Reply:
> Why would there be a trailing / ?

Why not? That's a valid path for a group `/foo/mygroup/`